README.md
<p align="center">
    <img src="docs/exodus.svg" alt="Exodus" height="142"/>
    <br>
    Exodus is a light, drop-in migration runner for Spring.
</p>

<p>
    <img id="badge--java" src="https://img.shields.io/badge/Java-17%2B-b07219" alt="Java17" />
    <img id="badge--spring" src="https://img.shields.io/badge/Spring-5%2B-6db33f" alt="test coverage" />
    <img id="badge--tests" src="https://img.shields.io/badge/tests-100%25%20%E2%9C%94-brightgreen" alt="test coverage" />
    <img id="badge--size" src="https://img.shields.io/badge/JAR%20size-~7%20kb-blueviolet" alt="size" />
    <img id="badge--version" src="https://img.shields.io/badge/version-1.2.0-white" alt="version" />
</p>

Exodus' aim is not to compete with incumbent migration runners on number of features, but rather to remove bloat and offer a simple solution to migrations in Spring applications.


## Use Exodus in a Spring project
1. Download the latest `exodus.jar` from [`dist/`](dist) above and place this JAR in your Spring project's `src/main/resources/lib/` directory.
2. Add Exodus as a dependency in your POM, taking care to set the version property as appropriate:
```
<dependency>
    <groupId>com.albertomh</groupId>
    <artifactId>exodus</artifactId>
    <version>1.1.0</version>
    <scope>system</scope>
    <systemPath>${basedir}/src/main/resources/lib/exodus-1.1.0.jar</systemPath>
</dependency>
```
3. Using the [sample application entrypoint](docs/SampleApplicationEntrypoint.java) as a guide, do the following:
    - Pass `scanBasePackages = {"com.albertomh.exodus"}` as a parameter to the `@SpringBootApplication` decorator.
    - Instantiate an application context with `SpringApplication.run(YourApplication.class, args)`.
    - Call `applicationContext.start()` inside `main()` to emit a `ContextStartedEvent`. This is the cue for Exodus to run any pending migrations.

These three steps are all that is needed to add Exodus to a project — you can now start writing migrations.


### Write migrations
Exodus will pick up any `.sql` files you place under `src/main/resources/db/migration/` in your Spring application.  
The following two best practices are recommended (but not enforced by Exodus):
- Subdivide `db/migration/` into directories named after the year the migrations they hold were written in.
- Have migrations follow the naming convention `YYYY-MM-DD_HH.MM__<MODULE>__<CHANGE>.sql` where `<MODULE>` is a subdivision of your app's functionality and `<CHANGE>` is a concise summary of the change enacted by the migration. For instance: `1970-01-01_09.00__auth__create-user.sql`.


### Sample output
The following line will appear in your Spring application's log the first time you run it after installing Exodus:
```
| exodus - Table `_schema_migration` has been created.
```
Exodus will create the table `_schema_migration` to keep track of the migrations that have been applied:
| id          | applied_at  | file_name      | checksum       |
| ----------- | ----------- | -------------- | -------------- | 
| `SERIAL PK` | `TIMESTAMP` | `VARCHAR(255)` | `VARCHAR(128)` |

If a valid migration is found within `db/migration/` (at any depth, see above) `_schema_migration` will be queried. If the migration is not listed in this table it will be applied and the following logged to the console:
```
| exodus - Migration `1970-01-01_09.00__auth__create-user.sql` has been applied.
```

After every run Exodus will log to the console a summary of operations taken: 
```
| exodus - Ignored [5] existing migrations. Applied [2] new migrations.
```

### The `MigrationCompleteEvent` event
After every run Exodus will publish a custom Spring event, the `MigrationCompleteEvent`.  
You can listen for this event in order to execute code after all migrations have been applied. Have your class implement `ApplicationListener<MigrationCompleteEvent>`, overriding the `onApplicationEvent` method: 
```
import org.springframework.context.ApplicationListener;
import com.albertomh.exodus.event.MigrationCompleteEvent;

public class MyComponent implements ApplicationListener<MigrationCompleteEvent> {
    @Override
    public void onApplicationEvent(MigrationCompleteEvent event) {
        // Your logic here.
    }
}
```

## Develop

### Build
Build Exodus with `./mvnw clean package`. This will create a JAR under `target/`.  
JARs should be placed under `dist/` for new releases — avoid doing this manually and instead run the `new_release.sh` script that takes care of this and other release-time tasks for you.


### Test
Tests are located under `src/test/` and laid out in the way common to Java projects, replicating the file structure of the application source code. 
Verify any changes you make by running `./mvnw clean test` (suite of unit & integration tests) from the project root.


### Cut a release
1. Merge all changes into the `main` branch and update the `<version>` property in the POM.
2. Run `new_release.sh`. This will run all tests, create a new JAR & place it in `dist/`, and update README badges.
3. Commit the new JAR and update `RELEASES.md`.
4. Tag the new release in git and push to origin.


## Project structure

### Dependencies
Exodus makes use of components in `org.springframework.core`, `o.s.context`, `o.s.util`, and `o.s.jdbc`.  
At compile-time these are provided by the `spring-boot-starter-web` & `spring-boot-starter-data-jpa` libraries.  
At run-time these are **not** provided as part of an uberJAR since Exodus is designed to be a dependency of a Spring application and as such will make use of the Spring libraries available in its run-time context.

In a test context two further dependencies — `spring-boot-starter-test` & the `H2` database — are required. The former provides testing libraries such as JUnit5, while the latter is an in-memory database used to quickly build and tear down databases during unit and integration testing.


### Entrypoint
#### MigrationRunner
- A datasource is injected into `MigrationRunner`'s constructor and this sets up a connection and statement for use throughout the Exodus instance's lifetime.
- `getMigrationScripts()` fetches migration scripts from the resources directory.
- `createSchemaMigrationTable()` creates the `_schema_migration` table if one does not already exist.
- `onApplicationEvent()` is Exodus' main loop, invoked when a `ContextStartedEvent` is emitted by the Spring application on startup. It will create the table in which migrations are recorded, loop through every migration script & assess whether it needs to be applied, and log appropriate messages to the console.


### Utilities
#### DatabaseUtils
A collection of utilities to interact with the database (more accurately, the `DataSource` injected into the `MigrationRunner`). Provides ways to inspect the tables in the database, list applied migrations, and apply a pending migration.


### Tests
#### TestingUtils
This module provides utilities used exclusively by the test suite.
- `createSchemaMigrationTable()` creates the `_schema_migration` table as part of setting up unit tests so that they don't depend on the same functionality implemented in the `MigrationRunner`.
- `addRowToSchemaMigrationTable()` simulates the behaviour of the runner by appending a row with arbitrary data to the `_schema_migration` table.

#### MigrationRunnerTest
A suite of tests for Exodus' main loop. A utility creates synthetic `ContextStartedEvent`s to simulate a Spring application's starting up.  
Unit tests check methods that fetch migration scripts from the resources directory or create the `_schema_migration` table. Integration tests assess the main loop's functionality under different conditions (blank database, previously-applied migrations, etc.).

#### DatabaseUtilsTest
A suite of tests for the `DatabaseUtils` tools. Unit tests verify simple functions such as counting the number of tables or listing applied migrations. Integration tests check the functionality of more complex procedures such as applying a migration.

**For both test suites:**
- The constructor sets up a new ephemeral H2 datasource, connection, and statement when a suite is instantiated.
- `beforeEach()` drops all objects in the H2 test database to make each test independent.


#### Test resources
Test resources are provided under `src/test/resources/db/migration/`, mimicking their location in a production application using Exodus. Two migrations are available to tests under the `1970/` subdirectory.


---

Copyright 2022 Alberto Morón Hernández  
This software is provided as open-source under the MIT License.


RELEASES.md
### 1.2.0
2022-07-13
- Improve robustness of `getMigrationScripts()` by ensuring picked up migration files are sorted as expected & using a more appropriate collection data type.

### 1.1.0
2022-07-03
- Emit a custom Spring Event (`MigrationCompleteEvent`) when Exodus succesfully completes a run.
- Improve management of database connections & statements by using `try-with-resources` statements. 

### 1.0.1
2022-06-25
- Make `MigrationRunner` public to allow use in application tests.

### 1.0.0
2022-05-20  
**MVP release**  
- Exodus integrates with Spring's application lifecycle as a drop-in dependency to run & manage database migrations.
- Checks for new migrations and applies them on application startup, keeping track of them in a dedicated table.
- Full test coverage: a suite of unit & integration tests covering database utilities and Exodus' main loop.
- Compiles to a lightweight JAR (~7 kb).

---

Copyright 2022 Alberto Morón Hernández


src/test/java/com/albertomh/exodus/TestingUtils.java
/**
 * Copyright 2022 Alberto Morón Hernández
 */
package com.albertomh.exodus;

import java.util.UUID;
import java.sql.SQLException;
import java.sql.Statement;


public class TestingUtils {

    public static String generateRandomString() {
        return UUID.randomUUID().toString().replace("-", "");
    }

    public static void createSchemaMigrationTable(Statement statement) {
        try {
            String tableSchema = """
                CREATE TABLE IF NOT EXISTS _schema_migration (
                    id SERIAL PRIMARY KEY,
                    applied_at TIMESTAMP WITH TIME ZONE DEFAULT NOW() NOT NULL,
                    file_name TEXT NOT NULL,
                    checksum TEXT NOT NULL
                );""";
            statement.execute(tableSchema);
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    public static void addRowToSchemaMigrationTable(Statement statement, String migrationFilename) {
        try {
            String tableSchema = String.format(
                "INSERT INTO _schema_migration (file_name, checksum) VALUES ('%s', '%s');",
                migrationFilename, generateRandomString()
            );
            statement.execute(tableSchema);
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    public static void addRowToSchemaMigrationTable(Statement statement) {
        addRowToSchemaMigrationTable(statement, generateRandomString());
    }

}


src/test/java/com/albertomh/exodus/MigrationRunnerTest.java
/**
 * Copyright 2022 Alberto Morón Hernández
 */
package com.albertomh.exodus;

import java.util.List;
import java.util.stream.Collectors;
import java.util.ArrayList;
import java.sql.Connection;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;
import javax.sql.DataSource;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestInstance;
import org.junit.jupiter.api.TestInstance.Lifecycle;
import org.junit.jupiter.api.BeforeEach;
import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;

import org.springframework.context.event.ContextStartedEvent;
import org.springframework.context.support.StaticApplicationContext;
import org.springframework.core.io.Resource;
import org.springframework.jdbc.datasource.embedded.EmbeddedDatabaseBuilder;
import org.springframework.jdbc.datasource.embedded.EmbeddedDatabaseType;
import org.slf4j.LoggerFactory;
import ch.qos.logback.classic.Level;
import ch.qos.logback.classic.Logger;
import ch.qos.logback.classic.spi.ILoggingEvent;
import ch.qos.logback.core.read.ListAppender;

import com.albertomh.exodus.event.MigrationCompleteEventPublisher;
import com.albertomh.exodus.util.DatabaseUtils;

@TestInstance(Lifecycle.PER_CLASS)
public class MigrationRunnerTest {

    private DataSource dataSource;
    private MigrationRunner runner;
    private MigrationCompleteEventPublisher migrationCompleteEventPublisher;

    Logger logger = (Logger) LoggerFactory.getLogger(MigrationRunner.class);
    List<ILoggingEvent> logList;

    public MigrationRunnerTest() {
        dataSource = new EmbeddedDatabaseBuilder().setType(EmbeddedDatabaseType.H2).build();
    }

    // ───── Utilities ─────────────────────────────────────────────────────────

    private ContextStartedEvent generateContextStartedEvent() {
        StaticApplicationContext staticApplicationContext = new StaticApplicationContext();
        return new ContextStartedEvent(staticApplicationContext);
    }

    // ───── Test lifecycle ────────────────────────────────────────────────────

    @BeforeEach
    private void setUp() {
        try (
            Connection conn = dataSource.getConnection();
            Statement statement = conn.createStatement();
        ) {
            statement.execute("DROP ALL OBJECTS;");
        } catch (SQLException e) {
            e.printStackTrace();
        }

        migrationCompleteEventPublisher = mock(MigrationCompleteEventPublisher.class);
        runner = new MigrationRunner(dataSource, migrationCompleteEventPublisher);

        logger.detachAndStopAllAppenders();
        ListAppender<ILoggingEvent> listAppender = new ListAppender<>();
        logger.addAppender(listAppender);
        listAppender.start();
        logList = listAppender.list;
    }

    // ───── Tests ─────────────────────────────────────────────────────────────

    @Test
    public void testGetMigrationScripts() {
        List<Resource> sqlScripts = MigrationRunner.getMigrationScripts();

        List<String> filenames = sqlScripts.stream().map(s -> s.getFilename()).collect(Collectors.toList());
        assertEquals("1970-01-01_09.00__auth__create-user.sql, 1970-02-01_10.00__email__create-email.sql", String.join(", ", filenames));
        assertEquals(2, sqlScripts.size());
    }

    @Test
    public void testCreatingTheSchemaMigrationTable() {
        try (
            Connection conn = dataSource.getConnection();
            Statement statement = conn.createStatement();
        ) {
            Integer initialTableCount = DatabaseUtils.countTables(statement);

            runner.createSchemaMigrationTable();

            assertEquals(0, initialTableCount);
            assertEquals(1, DatabaseUtils.countTables(statement));
            assertEquals(Level.INFO, logList.get(0).getLevel());
            assertEquals("exodus - Table `_schema_migration` has been created.", logList.get(0).getMessage());
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    @Test
    public void testCSETriggersMigrationsWithBlankDatabase() {
        try (
            Connection conn = dataSource.getConnection();
            Statement statement = conn.createStatement();
        ) {
            Integer initialTableCount = DatabaseUtils.countTables(statement);

            runner.onApplicationEvent(generateContextStartedEvent());

            assertEquals(0, initialTableCount);
            assertEquals(3, DatabaseUtils.countTables(statement));
            assertEquals(4, logList.size());
            assertEquals("exodus - Table `_schema_migration` has been created.", logList.get(0).getMessage());
            assertEquals("exodus - Migration `1970-01-01_09.00__auth__create-user.sql` has been applied.", logList.get(1).getMessage());
            assertEquals("exodus - Migration `1970-02-01_10.00__email__create-email.sql` has been applied.", logList.get(2).getMessage());
            String message = "Ignored [0] existing migrations. Applied [2] new migrations.";
            assertEquals(String.format("exodus - %s", message), logList.get(3).getMessage());
            verify(migrationCompleteEventPublisher).publishMigrationCompleteEvent(message);
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    @Test
    public void testCSETriggersMigrationsWithPopulatedDatabase() {
        try (
            Connection conn = dataSource.getConnection();
            Statement statement = conn.createStatement();
        ) {
            TestingUtils.createSchemaMigrationTable(statement);
            TestingUtils.addRowToSchemaMigrationTable(statement, "1970-01-01_09.00__auth__create-user.sql");
            Integer initialTableCount = DatabaseUtils.countTables(statement);

            runner.onApplicationEvent(generateContextStartedEvent());

            assertEquals(1, initialTableCount);
            assertEquals(2, DatabaseUtils.countTables(statement));
            assertEquals(2, logList.size());
            assertEquals("exodus - Migration `1970-02-01_10.00__email__create-email.sql` has been applied.", logList.get(0).getMessage());
            String message = "Ignored [1] existing migration. Applied [1] new migration.";
            assertEquals(String.format("exodus - %s", message), logList.get(1).getMessage());
            verify(migrationCompleteEventPublisher).publishMigrationCompleteEvent(message);
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    @Test
    public void testMigrationsAreRecordedInMigrationsTable() {
        try (
            Connection conn = dataSource.getConnection();
            Statement statement = conn.createStatement();
        ) {
            TestingUtils.createSchemaMigrationTable(statement);

            runner.onApplicationEvent(generateContextStartedEvent());

            ArrayList<String> migrationFilenames = new ArrayList<>();
            ArrayList<String> migrationChecksums = new ArrayList<>();
            ResultSet result = statement.executeQuery("SELECT * FROM _schema_migration;");
            while (result.next()) {
                migrationFilenames.add(result.getString("file_name"));
                migrationChecksums.add(result.getString("checksum"));
            }
            assertEquals("1970-01-01_09.00__auth__create-user.sql, 1970-02-01_10.00__email__create-email.sql", String.join(", ", migrationFilenames));
            assertEquals("F7CAE56B9AD4136BE9530CB25979F397", migrationChecksums.get(1));
        } catch (SQLException e) {
            logger.error(e.getMessage());
        }
    }

}

src/test/java/com/albertomh/exodus/util/DatabaseUtilsTest.java
/**
 * Copyright 2022 Alberto Morón Hernández
 */
package com.albertomh.exodus.util;

import java.sql.Connection;
import java.sql.SQLException;
import java.sql.Statement;
import javax.sql.DataSource;

import org.springframework.core.io.Resource;
import org.springframework.core.io.ClassPathResource;
import org.springframework.jdbc.datasource.embedded.EmbeddedDatabaseType;
import org.springframework.jdbc.datasource.embedded.EmbeddedDatabaseBuilder;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestInstance;
import org.junit.jupiter.api.TestInstance.Lifecycle;
import static org.junit.jupiter.api.Assertions.*;

import com.albertomh.exodus.TestingUtils;

@TestInstance(Lifecycle.PER_CLASS)
public class DatabaseUtilsTest {

    DataSource dataSource;

    public DatabaseUtilsTest() {
        dataSource = new EmbeddedDatabaseBuilder().setType(EmbeddedDatabaseType.H2).build();
    }

    // ───── Test lifecycle ────────────────────────────────────────────────────

    @BeforeEach
    private void beforeEach() {
        try (
            Connection conn = dataSource.getConnection();
            Statement statement = conn.createStatement();
        ) {
            statement.execute("DROP ALL OBJECTS;");
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    // ───── Tests ─────────────────────────────────────────────────────────────

    @Test
    public void testCountTablesWithEmptyDB() {
        try (
            Connection conn = dataSource.getConnection();
            Statement statement = conn.createStatement();
        ) {
            assertEquals(0, DatabaseUtils.countTables(statement));
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    @Test
    public void testCountTablesWithOneTable() {
        try (
            Connection conn = dataSource.getConnection();
            Statement statement = conn.createStatement();
        ) {
            TestingUtils.createSchemaMigrationTable(statement);

            assertEquals(1, DatabaseUtils.countTables(statement));
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    @Test
    public void testlistAppliedMigrationsForEmptyDB() {
        try (
            Connection conn = dataSource.getConnection();
            Statement statement = conn.createStatement();
        ) {
            TestingUtils.createSchemaMigrationTable(statement);

            assertEquals(0, DatabaseUtils.listAppliedMigrations(statement).size());
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    @Test
    public void testlistAppliedMigrationsAfterMigrationsRun() {
        try (
            Connection conn = dataSource.getConnection();
            Statement statement = conn.createStatement();
        ) {
            TestingUtils.createSchemaMigrationTable(statement);
            TestingUtils.addRowToSchemaMigrationTable(statement);
            TestingUtils.addRowToSchemaMigrationTable(statement);

            assertEquals(2, DatabaseUtils.listAppliedMigrations(statement).size());
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    @Test
    public void testApplyMigration() {
        try (
            Connection conn = dataSource.getConnection();
            Statement statement = conn.createStatement();
        ) {
            TestingUtils.createSchemaMigrationTable(statement);
            Integer initialTableCount = DatabaseUtils.countTables(statement);
            Integer initialAppliedMigrations = DatabaseUtils.listAppliedMigrations(statement).size();

            Resource migrationFile = new ClassPathResource("db/migration/1970/1970-01-01_09.00__auth__create-user.sql");
            DatabaseUtils.applyMigration(dataSource, migrationFile);

            assertEquals(1, initialTableCount);
            assertEquals(0, initialAppliedMigrations);
            assertEquals(2, DatabaseUtils.countTables(statement));
            assertEquals(1, DatabaseUtils.listAppliedMigrations(statement).size());
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

}


new_release.sh
#!/bin/bash

# Exodus
# Copyright 2022 Alberto Morón Hernández
# [github.com/albertomh/Exodus]
#
# New release
# ▔▔▔▔▔▔▔▔▔▔▔
# Run a full test suite and package into a JAR under `dist/`.
# Update badges in the README.md file.


SCRIPT_PATH=$(dirname $(realpath -s $0))
ROOT_DIR=$SCRIPT_PATH
# When invoked from a git hook, the context changes to the `.git/`
# directory. Use `pwd` instead when called by a git hook.
if [[ "$SCRIPT_PATH" == *".git"* ]]; then
    ROOT_DIR=$(pwd)
fi

get_version() {
    local pom_path="$ROOT_DIR/pom.xml"
    local pom_text=$(cat $pom_path)
    local version_re="<artifactId>exodus</artifactId>\s+<version>([0-9]\.[0-9]\.[0-9])</version>"
    if [[ $pom_text =~ $version_re ]]; then
        VERSION=${BASH_REMATCH[1]}
    fi
}

package_release() {
    printf "\nPackaging into 'dist/exodus-$VERSION.jar'\n\n"
    ./mvnw --quiet clean package
    cp ./target/*.jar ./dist
}

# Badges:  version
set_badges_in_readme() {
    printf "\nSetting badges in README.\n"
    local readme_path="$ROOT_DIR/README.md"

    # Set version badge.
    local version_badge_re='\/badge\/version.+?white'
    local new_version_badge="\/badge\/version-${VERSION}-white"
    sed -i -E "s/$version_badge_re/$new_version_badge/g" "$readme_path"
    printf "    Set badge ( version | ${VERSION} )\n"
}

cleanup() {
    printf "\nRemoving '/target'\n"
    rm -rf "$ROOT_DIR/target/"
}

# ─────────────────────────────────────────────────────────────────────────────

VERSION=""

main() {
    printf "\n──────────── BEGIN NEW_RELEASE SCRIPT ─────────────\n"

    get_version

    package_release

    set_badges_in_readme

    cleanup

    printf "\n─────────────────────── END ───────────────────────\n\n"
}
main

src/main/java/com/albertomh/exodus/util/DatabaseUtils.java
/**
 * Copyright 2022 Alberto Morón Hernández
 */
package com.albertomh.exodus.util;

import java.util.ArrayList;

import javax.sql.DataSource;

import java.io.IOException;
import java.io.InputStream;
import java.sql.Connection;
import java.sql.Statement;
import java.sql.ResultSet;
import java.sql.SQLException;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.core.io.Resource;
import org.springframework.jdbc.datasource.init.ScriptUtils;
import org.springframework.util.DigestUtils;

public final class DatabaseUtils {

    private static final Logger logger = LoggerFactory.getLogger(DatabaseUtils.class);

    private DatabaseUtils() { }

    /**
     * Count the number of tables in the 'public' schema.
     * @param statement
     * @return A count of existing tables.
     */
    public static Integer countTables(Statement statement) {
        Integer tableCount = 0;

        try {
            String tableCountSQL = """
                    SELECT count(*) AS \"table_count\"
                    FROM information_schema.tables
                    WHERE LOWER(table_schema) = 'public';""";
            ResultSet result = statement.executeQuery(tableCountSQL);
            while (result.next()) {
                tableCount = result.getInt("table_count");
            }
        } catch (SQLException e) {
            logger.error(e.getMessage());
        }

        return tableCount;
    }

    /**
     * @return An array of filenames of migrations that have been applied.
     */
    public static ArrayList<String> listAppliedMigrations(Statement statement) {
        ArrayList<String> appliedMigrations = new ArrayList<>();

        Integer tableCount = DatabaseUtils.countTables(statement);

        if (tableCount > 0) {
            try {
                String existingMigrationsSQL = "SELECT m.file_name FROM _schema_migration m;";
                ResultSet existingMigrations = statement.executeQuery(existingMigrationsSQL);
                while (existingMigrations.next()) {
                    appliedMigrations.add(existingMigrations.getString("file_name"));
                }
            } catch (SQLException e) {
                logger.error(e.getMessage());
            }
        }

        return appliedMigrations;
    }

    /**
     * Apply the given migration, recording this in `_schema_migration`.
     *
     * @param dataSource
     * @param script The migration SQL script to apply.
     */
    public static void applyMigration(
        DataSource dataSource,
        Resource script
        ) {
        // Generate an MD5 digest to uniquely identify each migration script.
        String scriptDigest = "";
        try {
            InputStream fileStream = script.getInputStream();
            byte[] scriptBytes = fileStream.readAllBytes();
            scriptDigest = DigestUtils.md5DigestAsHex(scriptBytes).toUpperCase();
        } catch (IOException e) {
            e.printStackTrace();
        }

        // Apply the migration script and update `_schema_migration`.
        try (
            Connection conn = dataSource.getConnection();
            Statement statement = conn.createStatement();
        ) {
            ScriptUtils.executeSqlScript(conn, script);

            String updateSQL = "INSERT INTO _schema_migration(file_name, checksum) "
                    .concat(String.format("VALUES ('%s','%s');", script.getFilename(), scriptDigest));
            statement.executeUpdate(updateSQL);

        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

}


src/main/java/com/albertomh/exodus/MigrationRunner.java
/**
 * Copyright 2022 Alberto Morón Hernández
 */
package com.albertomh.exodus;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.io.IOException;
import java.sql.Connection;
import java.sql.Statement;
import javax.sql.DataSource;
import java.sql.SQLException;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;
import org.springframework.context.ApplicationListener;
import org.springframework.context.event.ContextStartedEvent;
import org.springframework.core.io.Resource;
import org.springframework.core.io.support.ResourcePatternResolver;
import org.springframework.core.io.support.PathMatchingResourcePatternResolver;

import com.albertomh.exodus.event.MigrationCompleteEventPublisher;
import com.albertomh.exodus.util.DatabaseUtils;

@Component
public class MigrationRunner implements ApplicationListener<ContextStartedEvent> {

    Logger logger = LoggerFactory.getLogger(MigrationRunner.class);

    private DataSource dataSource;
    private MigrationCompleteEventPublisher migrationCompleteEventPublisher;

    public MigrationRunner(
        DataSource dataSource,
        MigrationCompleteEventPublisher migrationCompleteEventPublisher
        ) {
        this.dataSource = dataSource;
        this.migrationCompleteEventPublisher = migrationCompleteEventPublisher;
    }

    /**
     * Fetch all SQL scripts under `/db/migration/` and sort by filename.
     *
     * @return An array of Resources, each of them a script.
     */
    public static List<Resource> getMigrationScripts() {
        List<Resource> sqlScripts = new ArrayList<>();
        try {
            ClassLoader cl = MigrationRunner.class.getClassLoader();
            ResourcePatternResolver resolver = new PathMatchingResourcePatternResolver(cl);

            sqlScripts = Arrays.asList(resolver.getResources("classpath:/db/migration/**/*.sql"));
            Collections.sort(sqlScripts, Comparator.comparing(Resource::getFilename));

        } catch (IOException e) {
            e.printStackTrace();
        }

        return sqlScripts;
    }

    /**
     * Create the `_schema_migration` table if it does not yet exist.
     */
    public void createSchemaMigrationTable() {
        try (
            Connection conn = dataSource.getConnection();
            Statement statement = conn.createStatement();
        ) {
            String createMigrationsTableSQL = """
                CREATE TABLE IF NOT EXISTS _schema_migration (
                    id SERIAL PRIMARY KEY,
                    applied_at TIMESTAMP WITH TIME ZONE DEFAULT NOW() NOT NULL,
                    file_name VARCHAR(255) NOT NULL UNIQUE,
                    checksum VARCHAR(128) NOT NULL
                );
                """;
            statement.execute(createMigrationsTableSQL);
            logger.info("exodus - Table `_schema_migration` has been created.");
        } catch (SQLException e) {
            logger.error(e.getMessage());
        }
    }

    /**
     * Find any unapplied migrations on start-up and apply them.
     * 
     * @param event
     */
    @Override
    public void onApplicationEvent(ContextStartedEvent event) {
        // Initialise the `_schema_migration` table the first time Exodus runs.
        try (
            Connection conn = dataSource.getConnection();
            Statement statement = conn.createStatement();
        ) {
            Integer tableCount = DatabaseUtils.countTables(statement);
            if (tableCount == 0) {
                createSchemaMigrationTable();
            }
        } catch (SQLException e) {
            logger.error(e.getMessage());
        }

        List<Resource> sqlScripts = getMigrationScripts();

        // Loop over every SQL script and apply those that have not already been applied.
        Integer existingMigrationsCount = 0;
        Integer newMigrationsCount = 0;
        if (sqlScripts != null && dataSource != null) {
            ArrayList<String> appliedMigrations = new ArrayList<>();
            try (
                Connection conn = dataSource.getConnection();
                Statement statement = conn.createStatement();
            ) {
                appliedMigrations = DatabaseUtils.listAppliedMigrations(statement);
            } catch (SQLException e) {
                logger.error(e.getMessage());
            }

            for (Resource script : sqlScripts) {
                if (appliedMigrations == null
                    || appliedMigrations.size() == 0
                    || (appliedMigrations.size() > 0 && !appliedMigrations.contains(script.getFilename()))) {
                    DatabaseUtils.applyMigration(dataSource, script);
                    newMigrationsCount++;
                    logger.info(String.format("exodus - Migration `%s` has been applied.", script.getFilename()));
                } else {
                    existingMigrationsCount++;
                }
            }
        }

        try (
            Connection conn = dataSource.getConnection();
            Statement statement = conn.createStatement();
        ) {
            String existingMigPhrase = existingMigrationsCount == 1
                ? "existing migration"
                : "existing migrations";
            String newMigPhrase = newMigrationsCount == 1
                ? "new migration"
                : "new migrations";
            String message = String.format("Ignored [%d] %s. Applied [%d] %s.",
                existingMigrationsCount, existingMigPhrase,
                newMigrationsCount, newMigPhrase);
            logger.info(String.format("exodus - %s", message));

            migrationCompleteEventPublisher.publishMigrationCompleteEvent(message);

        } catch (SQLException e) {
            logger.error(e.getMessage());
        }
    }

}
